### ISSUE TITLE
API for custom metric reporting?

### ISSUE THREAD

**Issue by garrett361** [pytorch/torchtitan#1576]
It would be nice if it were easier to report custom metrics for particular models more easily, but currently this seems to require changing `train.py` and/or modifying `MetricsProcessor` in some invasive way.

Could we introduce an easier mechanism for reporting additional metrics for specific models? A specific use case is to log EP token routing metrics, like [shown here](https://github.com/pytorch/torchtitan/issues/1467#issuecomment-3130249678) and whose custom implementation is [here](https://github.com/rakkit/torchtitan/blob/95732cac15e3c48983328961210b9e0b61e02b1d/torchtitan/train.py?plain=1#L581-L585).  I also sometimes want to track activation magnitudes at various layers.

One idea I had is to leverage the `extra_metrics` arg of the `MetricsProcessor.log` method, which is currently only used to log the lr and number of toks seen:
https://github.com/pytorch/torchtitan/blob/6fc499f6f5b32151a799188be2208cfb09faed30/torchtitan/train.py?plain=1#L517-L527

We could do something like give `ModelProtocol` a `get_extra_metrics` method:
```py
class ModelProtocol(Protocol):
    [...]
    def get_extra_metrics(self) -> None | dict:
        return None
```

and modify the reporting code to something like:
```py
extra_metrics = {
    "n_tokens_seen": global_ntokens_seen,
    "lr": lr,
}
custom_metrics = [mp.get_custom_metrics() for mp in self.model_parts]
for cm in custom_metrics:
    if cm is not None:
        extra_metrics.update(custom_metrics)
self.metrics_processor.log(
    self.step,
    global_avg_loss,
    global_max_loss,
    grad_norm.item(),
    extra_metrics=extra_metrics,
)
```
This can get a bit confusing in complex PP cases, but it's a start. 

Thoughts? CC @tianyu-l @rakkit

**Comment by rakkit** [comment]
looks good. One needs to make sure the metrics in `extra_metrics` are already reduced.

**Comment by garrett361** [comment]
yeah, it would be up to the user to add all proper reduction logic for their setup to `get_extra_metrics`. So I guess a better signature would be:
```py
def get_extra_metrics(self, parallel_dims: None | ParallelDims = None) -> None | dict:
        return None
```
I'll put up a PR. Might throw in a `*args, **kwargs` for extensibility, too.

**Comment by rakkit** [comment]
cool thanks!   For moe, if we add ep_usage, does it also make sense we also add the experts' routing entropy? (We only need to add this in moe.py as all the moe model ends up there)

**Comment by garrett361** [comment]
Let's have a separate discussion on exactly what metrics we might want to track (though I'd be interested in tracking routing entropy myself), since any metrics have a perf costs and I want to just get the plumbing set up first.
